test/Microsoft.ML.Tests/TextLoaderTests.cs

```diff
@@ -840,6 +840,302 @@ public void TestTextLoaderBackCompat_VerWritt_0x0001000C()
+
+            // Load values from iris.txt
+            DataViewSchema columnsPeriod = textDataDecimalMarkerPeriod.Schema;
+            using DataViewRowCursor cursorPeriod = textDataDecimalMarkerPeriod.GetRowCursor(columnsPeriod);
+            UInt32 labelPeriod = default;
+            ValueGetter<UInt32> labelDelegatePeriod = cursorPeriod.GetGetter<UInt32>(columnsPeriod[0]);
+            VBuffer<Single> featuresPeriod = default;
+            ValueGetter<VBuffer<Single>> featuresDelegatePeriod = cursorPeriod.GetGetter<VBuffer<Single>>(columnsPeriod[1]);
+
+            // Iterate over each row and save labels and features to array for future comparison
+            int count = 0;
+            while (cursorPeriod.MoveNext())
+            {
+                //Get values from respective columns
+                labelDelegatePeriod(ref labelPeriod);
+                featuresDelegatePeriod(ref featuresPeriod);
+                labels[count] = labelPeriod;
+                features[count] = featuresPeriod.GetValues().ToArray();
+                count++;
+            }
+
+            // Read dataset with comma as decimal marker.
+            string dataPathDecimalMarkerComma = GetDataPath("iris-decimal-marker-as-comma.txt");
+            var readerDecimalMarkerComma = new TextLoader(mlContext, new TextLoader.Options()
+            {
+                Columns = new[]
+                        {
+                            new TextLoader.Column("Label", DataKind.UInt32, 0),
+                            new TextLoader.Column("Features", DataKind.Single, new [] { new TextLoader.Range(1, 4) }),
+                        },
+                DecimalMarker = ','
+            });
+            var textDataDecimalMarkerComma = readerDecimalMarkerComma.Load(GetDataPath(dataPathDecimalMarkerComma));
+
+            // Load values from iris-decimal-marker-as-comma.txt
+            DataViewSchema columnsComma = textDataDecimalMarkerComma.Schema;
+            using DataViewRowCursor cursorComma = textDataDecimalMarkerComma.GetRowCursor(columnsComma);
+            UInt32 labelComma = default;
+            ValueGetter<UInt32> labelDelegateComma = cursorComma.GetGetter<UInt32>(columnsComma[0]);
+            VBuffer<Single> featuresComma = default;
+            ValueGetter<VBuffer<Single>> featuresDelegateComma = cursorComma.GetGetter<VBuffer<Single>>(columnsComma[1]);
+
+            // Check values from iris-decimal-marker-as-comma.txt match those in iris.txt
+            count = 0;
+            while (cursorComma.MoveNext())
+            {
+                //Get values from respective columns
+                labelDelegateComma(ref labelComma);
+                featuresDelegateComma(ref featuresComma);
+                Assert.Equal(labels[count], labelComma);
+                Assert.Equal(features[count], featuresComma.GetValues().ToArray());
+                count++;
+            }
+        }
+
+        [Fact]
+        public void TestCommaAsDecimalMarkerDouble()
```

Comment:
This test is very similar to `TestCommaAsDecimalMarkerFloat()` above. This is because it is necessary to test decimal markers with floats **and** doubles. I was wishing to use `InlineData` to input a bool for using doubles and floats, however the specification of float/double is necessary to instantiate the VBuffer and ValueGetters, and there is no better way to specify varying types that I know.

Reply:
As discussed offline, I agree with your approach of having 2 separate tests for each case, for the reasons you've mentioned above.

Reply:
You can do this a bit better by refactoring the core test as `TestCommaAsDecimalMarker<T>()` and then calling `TestCommaAsDecimalMarker<float>` from within `TestCommaAsDecimalMarkerFloat()`. You will still have two tests, but the code will be easier to follow that this is the same test on two different types.


---
In reply to: [429001199](https://github.com/dotnet/machinelearning/pull/5145#discussion_r429001199) [](ancestors = 429001199)

Reply:
@harishsk now that in `TestCommaAsDecimalMarkerFloat()` I am testing both the .csv (with comma as both separator and decimal marker) and the .txt versions of the same dataset by using `InlineDataAttribute`, I think both tests would become harder to follow with the addition of a generic type parameter and calling `TestCommaAsDecimalMarker<float>` from within `TestCommaAsDecimalMarkerFloat()`

Reply:
Please try it out. This is the pattern used in a lot of onnx conversion tests and it has improved readability there.

Reply:
@harishsk thank you for your suggestion. I have now compressed the two tests, where I have a theory `TestCommaAsDecimalMarker(bool useCsvVersion)` test that calls a `TestCommaAsDecimalMarkerHelper<T>(bool useCsvVersion)` helper test. If only I knew a way to pass in the floating point value types `float` and `double` as inputs to `TestCommaAsDecimalMarkerHelper<T>(bool useCsvVersion)`, I would have done the helper test this way to only have 1 unit test:

```csharp
    [Theory]
    [InlineData(true, typeof(float))]
    [InlineData(true, typeof(double))]
    [InlineData(false, typeof(float))]
    [InlineData(false, typeof(double))]
    public void TestCommaAsDecimalMarkerHelper(bool useCsvVersion, GENERIC_TYPE featureType)
    {
        ...
    }